File: src/media_advisor/pipeline/summarizer.py
"""AI-based summary generator for a single video.

Porting of src/pipeline/summarizer.ts.
"""

from media_advisor.models.claims import Claim

_SUMMARIZE_SYSTEM = """Sei un analista di media sportivi italiani.
Ti viene dato il testo di un video di commento calcistico e i claim principali estratti.
Produci un riassunto in italiano di 2-3 frasi che:
1. Descriva il tema centrale del video
2. Sintetizzi la posizione principale dell'autore
3. Menzioni eventuali ipotesi o conclusioni chiave
Max 150 parole, in italiano."""


async def generate_summary(
    api_key: str,
    model: str,
    title: str | None,
    author: str | None,
    full_text: str,
    claims: list[Claim],
) -> str:
    claims_text = "\n".join(
        f"- [{c.dimension}] {c.claim_text}" for c in claims[:8]
    )
    user_content = ""
    if title:
        user_content += f"Titolo: {title}\n"
    if author:
        user_content += f"Autore: {author}\n"
    user_content += f"\nTesto (estratto):\n{full_text[:3000]}\n\nClaim principali:\n{claims_text}"

    import openai

    client = openai.AsyncOpenAI(api_key=api_key)
    completion = await client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": _SUMMARIZE_SYSTEM},
            {"role": "user", "content": user_content},
        ],
        max_tokens=200,
        temperature=0.3,
    )
    return (completion.choices[0].message.content or "").strip()

File: src/pipeline/summarizer.ts
/**
 * Step 2E — AI summary generation from full transcript text + claims.
 * Produces a 2-3 sentence summary in Italian of the video's key points.
 */

import OpenAI from "openai";
import type { Claim } from "../schema/claims.js";

const SUMMARIZE_SYSTEM = `Sei un analista di media sportivi italiani. Ricevi la trascrizione di un video YouTube di un opinionista calcistico e una lista di claim estratti.

Produci un riassunto in italiano di 2-3 frasi che:
1. Descriva il tema centrale del video (cosa viene discusso/analizzato)
2. Sintetizzi la posizione principale dell'autore
3. Menzioni eventuali ipotesi o conclusioni chiave

Il riassunto deve essere:
- Oggettivo e basato sul contenuto
- Concreto (nomi, fatti, posizioni specifiche)
- Max 150 parole
- In italiano`;

export async function generateSummary(
  openai: OpenAI,
  input: {
    title?: string;
    author?: string;
    fullText: string;
    claims: Claim[];
  }
): Promise<string> {
  const claimsText = input.claims
    .slice(0, 8)
    .map((c) => `- ${c.claim_text}`)
    .join("\n");

  const truncatedText = input.fullText.slice(0, 3000);

  let userContent = "";
  if (input.title) userContent += `Titolo: ${input.title}\n`;
  if (input.author) userContent += `Autore: ${input.author}\n\n`;
  userContent += `Trascrizione (estratto):\n${truncatedText}\n\n`;
  if (claimsText) userContent += `Claim estratti:\n${claimsText}\n`;

  const completion = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [
      { role: "system", content: SUMMARIZE_SYSTEM },
      { role: "user", content: userContent },
    ],
    max_tokens: 200,
    temperature: 0.3,
  });

  return completion.choices[0]?.message?.content?.trim() ?? input.title ?? "Video";
}
